### DISCUSSION TITLE
Algorithme de recommandation

### DISCUSSION THREAD

**Discussion by pierrbt** [Mysterious-Developers/InfinIdea#44] [General]
Bienvenue sur cette nouvelle discussion.
je l'ouvre pour réfléchir sur le sujet : 
> Comment recommander des articles aux utilisateurs.

Pour information, nous disposons du nombre de vues, du nombre de likes, aussi aussi exactement qui a regardé/liké chaque article, on a aussi des hashtags qui permettent de juger la catégorie de l'article.

Pour utiliser ça, on a besoin d'une idée de comment ça va marcher avec par ex une formule qui calcule la chance qu'un article corresponde à l'utilisateur, et ensuite le mettre en pratique avec un script, on peut utiliser des langages comme PHP, Python, ou autre et qui communiquerait avec la base de donnée MySQL.

Toutes les idées sont bienvenues, merci !
Pierre

**Comment by pierrbt** [comment]
On peut aussi savoir quand l'article a été posté et je vais faire un système pour pouvoir suivre les auteurs d'article, je vais aussi faire en sorte de demander à l'utilisateur ce qu'il préfère lors de l'inscription et je vais stocker les tags qui étaient présents sur un article que l'utilisateur a lu, comme ça on saura quel type d'article il lit,
ça commence à faire pas mal de données 😅

**Comment by olarno** [comment]
Avez vous un schéma de base de donnée ? 

en partant d'un utilisateur, pouvons nous retourner le nombre d'article vues ? Avez les types on pourrait faire ressortir un % en fonction des lectures, et du coup proposer des articles composés avec les même tags. 

Example : user 1 a lu 15 articles [ 10 - dev, 5 - php] ce qui donne un ratio de 70% de dev 30% php , on peut donc lui proposer les nouveau article portant sur le dev.

cela reste theorique mais on peut faire un POC de l'algo

**Comment by pierrbt** [comment]
@olarno 

> Avez vous un schéma de base de donnée ?

Je n'en avais pas fait mais je viens d'en créer un avec les tables qui peuvent être utiles pour les suggestions :

![image](https://user-images.githubusercontent.com/83456236/200119173-df44c9ff-d3e6-4568-bc0b-6bf29950f465.png)

Pour information, la base de données est une MySQL (plus précisément MariaDB).

---

> en partant d'un utilisateur, pouvons nous retourner le nombre d'article vues ?

Oui avec la table `views` il est facile de retrouver les articles vus par l'utilisateur, si on veut récupérer leur id : (aid : id de l'article, uid : id de l'utilisateur)
```sql
SELECT DISTINCT aid FROM views WHERE uid = 10 ORDER BY date DESC LIMIT 20;
```

---

> Avez les types on pourrait faire ressortir un % en fonction des lectures, et du coup proposer des articles composés avec les même tags.

Oui ce serait une bonne idée mais la majorité des articles contiendront surement un #dev donc il faudrait essayer de trouver des articles non-lus, qui auraient une majorité des tags en commun.

Si ça t'intéresse, on peut faire un POC avec toutes les informations. Je pense que ça tiendrait compte des 20 derniers articles lus comme ça ça suivrait la tendance de l'utilisateur assez rapidement, et aussi je pense mettre en avant les articles qui ont beaucoup de vues/likes ou un bon rapport likes/vues. Dit moi ce que tu en penses.

Pierre

**Reply by olarno** [reply]
> Oui ce serait une bonne idée mais la majorité des articles contiendront surement un #dev donc il faudrait essayer de trouver des articles non-lus, qui auraient une majorité des tags en commun.

On pourrait avoir une table de `tag` liée aux `articles`  de la on peut ressortir un comptage des articles/vues/tag et donc des user/article/vue/tag afin de permettre de faire un pourcentage. Je vais réfléchir a un schéma pour ça. 


Pour faire une POC, je te renvoie sur la proposition [#53](https://github.com/Mysterious-Developers/InfinIdea/issues/53) afin de pouvoir mettre cela en place.

**Reply by pierrbt** [reply]
C'est vraiment sympa de t'investir autant 😁.
Les tags étaient pour l'instant mis dans une colonne de la table article mais ce n'est pas plus mal si c'est dans une table séparée pour les jointures etc...
Je vais faire une issue pour faire ça 😉
